Use an async reply for RemoteImageBufferSet::endPrepareForDisplay.

mattwoodrow opened the pull request:
#### e5215d63d0216296a4378d4d70a4888da093da73
<pre>
Use an async reply for RemoteImageBufferSet::endPrepareForDisplay.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301828">https://bugs.webkit.org/show_bug.cgi?id=301828</a>
&lt;<a href="https://rdar.apple.com/problem/163899986">rdar://problem/163899986</a>&gt;

Reviewed by Kimmo Kinnunen.

This currently uses a separate message (didPrepareForDisplay) as a reply, which
gets delivered to a work queue.

We should be able to use sendWithAsyncReplyOnDispatcher instead, to more closely
link the message and its reply.

This also removes local state in RemoteImageBufferSetProxy tracking the pending
reply message, which should simplify the future case where there are multiple in
flight.

Tests: Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp
       Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h
       Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::endPrepareForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::sendWithAsyncReplyOnDispatcher):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::sendWithAsyncReply):
(WebKit::RemoteImageBufferSetProxy::prepareToDisplay):
(WebKit::RemoteImageBufferSetProxy::close):
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
(WebKit::RemoteImageBufferSetProxy::willPrepareForDisplay):
(WebKit::RemoteImageBufferSetProxy::disconnect):
(WebKit::RemoteImageBufferSetProxy::didPrepareForDisplay): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
(WebKit::RemoteImageBufferSetProxy::generation const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.messages.in: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::create): Deleted.
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::beginShutdown): Deleted.
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::waitUntilShutdown): Deleted.
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::WorkQueueWithShutdown): Deleted.
(TestWebKitAPI::AutoWorkQueue::AutoWorkQueue): Deleted.
(TestWebKitAPI::AutoWorkQueue::queue): Deleted.
(TestWebKitAPI::AutoWorkQueue::~AutoWorkQueue): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::create):
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::beginShutdown):
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::waitUntilShutdown):
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::WorkQueueWithShutdown):
(TestWebKitAPI::AutoWorkQueue::AutoWorkQueue):
(TestWebKitAPI::AutoWorkQueue::queue):
(TestWebKitAPI::AutoWorkQueue::~AutoWorkQueue):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: <a href="https://commits.webkit.org/302557@main">https://commits.webkit.org/302557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2944075852c989f3fe6d0b76934e760c6bd9d986

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129319 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136696 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80712 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0456b7b3-56c5-4408-95d8-96decdcb32fa) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131190 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1452 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98486 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66389 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bd713dee-7790-44d2-8285-26c101f6fb56) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1176 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115830 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79139 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5113102b-e703-4546-94cd-6fd463fd24ca) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/128669 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1099 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33953 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79973 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109550 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139168 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1365 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1324 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107012 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1408 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112168 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106860 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27241 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1114 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30689 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54001 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1436 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64800 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1254 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1289 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1358 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
```diff
@@ -94,7 +93,7 @@ void RemoteImageBufferSet::updateConfiguration(const RemoteImageBufferSetConfigu
+void RemoteImageBufferSet::endPrepareForDisplay(RenderingUpdateID renderingUpdateID, CompletionHandler<void(ImageBufferSetPrepareBufferForDisplayOutputData, RenderingUpdateID)>&& completionHandler)
 {
     m_context.reset();
 
```
Completion handler not called for non-COCOA cases..

mattwoodrow replied:
That was the previous state too (except it compiled, obviously).

No real reason for it to be platform specific though.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
```diff
@@ -269,13 +243,25 @@ std::unique_ptr<ThreadSafeImageBufferSetFlusher> RemoteImageBufferSetProxy::flus
+        RELEASE_ASSERT(!isMainRunLoop());
+
+        BufferSetBackendHandle handle;
+        handle.bufferHandle = WTFMove(outputData.backendHandle);
```
This could have the cancel logic that was in disconnect.It's not clear to me from this code you end up with same call of that you had in disconnect case, `pendingFlush->setHandles({})`. If it happens so, then it's ok.


E.g. when async reply handler gets cancelled due to gpup crash, the args will get the cancel constructed objects, typically "empty" objects.

mattwoodrow replied:
It should be the same.

Both `ImageBufferSetPrepareBufferForDisplayOutputData` and `BufferSetBackendHandle` are just collections of `std::optional`s.

The old disconnect case default-constructed a `BufferSetBackendHandle`, the new one default constructs a `ImageBufferSetPrepareBufferForDisplayOutputData` and then converts it, but it should all be `std::nullopt` and identical.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -247,6 +249,62 @@ std::optional<StreamClientConnection::AsyncReplyID> StreamClientConnection::send
 }
 
+template<typename T, typename C, typename U, typename V, typename W>
+std::optional<StreamClientConnection::AsyncReplyID> StreamClientConnection::sendWithAsyncReplyOnDispatcher(T&& message, GuaranteedSerialFunctionDispatcher& dispatcher, C&& completionHandler, ObjectIdentifierGeneric<U, V, W> destinationID, Timeout timeout)
```
Would you be able to add a test for this in StreamConnectionTests.cpp ?

mattwoodrow replied:
Done

mattwoodrow committed:
Use an async reply for RemoteImageBufferSet::endPrepareForDisplay.
https://bugs.webkit.org/show_bug.cgi?id=301828
<rdar://problem/163899986>

Reviewed by Kimmo Kinnunen.

This currently uses a separate message (didPrepareForDisplay) as a reply, which
gets delivered to a work queue.

We should be able to use sendWithAsyncReplyOnDispatcher instead, to more closely
link the message and its reply.

This also removes local state in RemoteImageBufferSetProxy tracking the pending
reply message, which should simplify the future case where there are multiple in
flight.

Tests: Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp
       Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h
       Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::endPrepareForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::sendWithAsyncReplyOnDispatcher):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/PrepareBackingStoreBuffersData.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::sendWithAsyncReply):
(WebKit::RemoteImageBufferSetProxy::prepareToDisplay):
(WebKit::RemoteImageBufferSetProxy::close):
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
(WebKit::RemoteImageBufferSetProxy::willPrepareForDisplay):
(WebKit::RemoteImageBufferSetProxy::disconnect):
(WebKit::RemoteImageBufferSetProxy::didPrepareForDisplay): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
(WebKit::RemoteImageBufferSetProxy::generation const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.messages.in: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::create): Deleted.
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::beginShutdown): Deleted.
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::waitUntilShutdown): Deleted.
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::WorkQueueWithShutdown): Deleted.
(TestWebKitAPI::AutoWorkQueue::AutoWorkQueue): Deleted.
(TestWebKitAPI::AutoWorkQueue::queue): Deleted.
(TestWebKitAPI::AutoWorkQueue::~AutoWorkQueue): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::create):
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::beginShutdown):
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::waitUntilShutdown):
(TestWebKitAPI::AutoWorkQueue::WorkQueueWithShutdown::WorkQueueWithShutdown):
(TestWebKitAPI::AutoWorkQueue::AutoWorkQueue):
(TestWebKitAPI::AutoWorkQueue::queue):
(TestWebKitAPI::AutoWorkQueue::~AutoWorkQueue):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: https://commits.webkit.org/302557@main
